x/wasm/relay_pingpong_test.go

```diff
@@ -121,6 +120,189 @@ func TestPinPong(t *testing.T) {
 
 }
 
+func TestPinPongWithAppLevelError(t *testing.T) {
+	var (
+		coordinator = ibc_testing.NewCoordinator(t, 2)
+		chainA      = coordinator.GetChain(ibc_testing.GetChainID(0))
+		chainB      = coordinator.GetChain(ibc_testing.GetChainID(1))
+	)
+	_ = chainB.NewRandomContractInstance() // skip 1 id
+	var (
+		pingContractAddr = chainA.NewRandomContractInstance()
+		pongContractAddr = chainB.NewRandomContractInstance()
+	)
+	require.NotEqual(t, pingContractAddr, pongContractAddr)
+
+	pingContract := &player{t: t, actor: ping, chain: chainA, contractAddr: pingContractAddr}
+	pongContract := &player{t: t, actor: pong, chain: chainB, contractAddr: pongContractAddr}
+
+	wasmkeeper.MockContracts[pingContractAddr.String()] = pingContract
+	wasmkeeper.MockContracts[pongContractAddr.String()] = pongContract
+
+	var (
+		sourcePortID       = wasmkeeper.PortIDForContract(pingContractAddr)
+		counterpartyPortID = wasmkeeper.PortIDForContract(pongContractAddr)
+	)
+	clientA, clientB, connA, connB := coordinator.SetupClientConnections(chainA, chainB, clientexported.Tendermint)
+	connA.NextChannelVersion = ping
+	connB.NextChannelVersion = pong
+
+	channelA, channelB := coordinator.CreateChannel(chainA, chainB, connA, connB, sourcePortID, counterpartyPortID, channeltypes.UNORDERED)
+	var err error
+
+	const startValue uint64 = 100
+	const rounds = 3
+	s := startGame{
+		ChannelID: channelA.ID,
+		Value:     startValue,
+		MaxValue:  rounds - 1,
+	}
+	startMsg := &wasm.MsgExecuteContract{
+		Sender:   chainA.SenderAccount.GetAddress(),
+		Contract: pingContractAddr,
+		Msg:      s.GetBytes(),
+	}
+	// send from chainA to chainB
+	err = coordinator.SendMsgs(chainA, chainB, clientB, startMsg)
+	require.NoError(t, err)
+
+	t.Log("Duplicate messages are due to check/deliver tx calls")
+
+	var (
+		activePlayer  = ping
+		pingBallValue = startValue
+	)
+	for i := 1; i <= rounds-1; i++ { // play some rounds before reaching max value
+		t.Logf("++ round: %d\n", i)
+		ball := NewHit(activePlayer, pingBallValue)
+
+		seq := uint64(i)
+		pkg := channeltypes.NewPacket(ball.GetBytes(), seq, channelA.PortID, channelA.ID, channelB.PortID, channelB.ID, doNotTimeout, 0)
+		ack := ball.BuildAck()
+
+		err = coordinator.RelayPacket(chainA, chainB, clientA, clientB, pkg, ack.GetBytes())
```

Comment:
This is odd, you pass in the ack instead of it coming from the contract

Reply:
This is a "workaround" in ibc tests. The relayer should watch the chain instead and read the packet/ack information from the events exposed there. Unfortunately this is not implemented into the coordinator (yet) and also not a simple task that I could squeeze in easily.
This setup is also used in other ibc tests so we are in good company at least.

Reply:
Fair enough, limits of test environment.